File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxAncestorDiff = function(root) {
  let ans = 0;
  
  const dfs = (node, min = Infinity, max = 0) => {
    if (!node) return;
    
    min = Math.min(min, node.val);
    max = Math.max(max, node.val);
    
    dfs(node.left, min, max);
    dfs(node.right, min, max);
    
    ans = Math.max(ans, max - min);
  }
  
  dfs(root);
  
  return ans;
};

File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        ans = 0
        
        def dfs(node, max_val = 0, min_val = inf):
            nonlocal ans
            if not node:
                return 
              
            max_val, min_val = max(max_val, node.val), min(min_val, node.val)
            dfs(node.left, max_val, min_val)
            dfs(node.right, max_val, min_val)
            ans = max(ans, max_val - min_val)
            
            
        dfs(root)
        return ans